sw/simulator/nps/nps_fdm_gazebo.cpp

```diff
@@ -585,4 +624,130 @@ static void read_image(
+  int32_t range_sensors_orientation_int32[NPS_GAZEBO_RANGE_AMOUNT_SENSORS];
+
+  const double range_orientation[] = { NPS_GAZEBO_RANGE_ORIENTATION };
+  const double range_orientation_agl[] = { NPS_GAZEBO_RANGE_ORIENTATION_AGL };
+
+  uint8_t ray_sensor_count_selected = 0;
+  float range_sensor_down = 0;
+
+  //Loop through all ray sensors found in gazebo
+  for (int i = 0; i < ray_sensor_count_in_gazebo; i++) {
+
+    //Read out the pose from per ray sensors in gazebo
+    ignition::math::Pose3d pose3d_sensor = RaySensorPtr_array[i]->Pose();
+    gazebo::math::Pose pose_sensor = gazebo::math::Pose(pose3d_sensor);
+
+    /* Check the orientations of the ray sensors found in gazebo, if they are similar (within 5 deg) to the orientations
+     * given in the airframe file in NPS_GAZEBO_RANGE_ORIENTATION
+     * NOTE: In gazebo, the y and z axis are inverted compared to paparazzi
+     */
+    bool found_a_match = false;
+    int8_t index_of_raysensor = 0;
+    for (int k = 0; k < NPS_GAZEBO_RANGE_AMOUNT_SENSORS; k++) {
+      if (RadOfDeg(5) > fabs((float)pose_sensor.rot.GetRoll() - range_orientation[k * 3 + 0]) &&
+          RadOfDeg(5) > fabs((float)(-1 * pose_sensor.rot.GetPitch()) - range_orientation[k * 3 + 1]) &&
+          RadOfDeg(5) > fabs((float)(-1 * pose_sensor.rot.GetYaw()) - range_orientation[k * 3 + 2])) {
+        found_a_match = true;
+        index_of_raysensor = k;
+        break;
+      }
+    }
+
+    /* If the orientations in the gazebo simulation are the same as the ones given in the airframe file
+     *  copy the values of the orientation in the array, in the order that the orientation are indicated
+     *   in the airframe file
+     */
+    if (found_a_match) {
+      range_sensors_uint16[index_of_raysensor] = (uint16_t)(RaySensorPtr_array[i]->Range(0) * 1000.);
+      if (range_sensors_uint16[index_of_raysensor] == 0 || isinf(range_sensors_uint16[index_of_raysensor])) {
+        range_sensors_uint16[index_of_raysensor] = UINT16_MAX;
+      }
+      for (int n = 0; n < 3; n++) {
+        range_sensors_orientation_int32[index_of_raysensor * 3 + n] =
+          ANGLE_BFP_OF_REAL(range_orientation[index_of_raysensor * 3 + n]);
+      }
+      ray_sensor_count_selected++;
+    }
+
+#ifdef NPS_GAZEBO_RANGE_ORIENTATION_AGL
+    /* One of the range sensors can be selected to act like a sonar. It will check which of
+     * the ray sensors is looking downwards (or any orientation you specify), within 5 degrees
+     * of variation, and select that as the sonar.
+     */
+    if (RaySensorPtr_array[i]->Range(0) != 0 && !isinf(RaySensorPtr_array[i]->Range(0)) &&
+        RadOfDeg(5) > fabs((float)pose_sensor.rot.GetRoll() - range_orientation_agl[0]) &&
+        RadOfDeg(5) > fabs((float)(-1 * pose_sensor.rot.GetPitch()) - range_orientation_agl[1]) &&
+        RadOfDeg(5) > fabs((float)(-1 * pose_sensor.rot.GetYaw()) - range_orientation_agl[2])) {
+      range_sensor_down = RaySensorPtr_array[i]->Range(0);
```

Comment:
alg instead of range_sensor_down. Also agl should be from the angle looking down not from the angle_orientation_agl matrix.

Reply:
changing the name to alg I agree. I would like to keep the range_orientation_agl since it seems less hardcoded and clean, so at least it is clear what the downward angle is, which is defined on top of the this file.